### DISCUSSION TITLE
How to exclude QT windows with client decorations?

### DISCUSSION THREAD

**Discussion by arielnmz** [yshui/picom#944] [Q&A]
Specifically in the case of Telegram (but I'm sure other QT based applications might have this issue) I see that the menus and other dialogs have this awful blurred space around them:

![image](https://user-images.githubusercontent.com/24735547/85956724-98be3500-b93c-11ea-88a2-8c54bdb3638f.png)

How do we tell compton to not blur and put shadows on them? I was looking for something similar to `_GTK_FRAME_EXTENTS` but either I can't figure out the right language to look for it or such a thing simply doesn't exist on QT.

This is what `xprop` can tell me about some of Telegram menus:

```
_NET_WM_USER_TIME(CARDINAL) = 6261168
WM_TRANSIENT_FOR(WINDOW): window id # 0x1c00006
_NET_WM_ICON_NAME(UTF8_STRING) =
_NET_WM_ICON(CARDINAL) = 	Icon (64 x 64):

[icon excluded]
    
XdndAware(ATOM) = BITMAP
WM_NAME(STRING) =
_NET_WM_NAME(UTF8_STRING) = "TelegramDesktop"
_MOTIF_WM_HINTS(_MOTIF_WM_HINTS) = 0x2, 0x1, 0x0, 0x0, 0x0
_NET_WM_WINDOW_TYPE(ATOM) = _KDE_NET_WM_WINDOW_TYPE_OVERRIDE, _NET_WM_WINDOW_TYPE_NORMAL
_XEMBED_INFO(_XEMBED_INFO) = 0x0, 0x1
WM_CLIENT_LEADER(WINDOW): window id # 0x1c00008
WM_HINTS(WM_HINTS):
        Client accepts input or input focus: True
        window id # of group leader: 0x1c00008
WM_CLIENT_MACHINE(STRING) = "bc-1.bc"
_NET_WM_PID(CARDINAL) = 2
_NET_WM_SYNC_REQUEST_COUNTER(CARDINAL) = 29360442
WM_CLASS(STRING) = "telegram-desktop", "TelegramDesktop"
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, WM_TAKE_FOCUS, _NET_WM_PING, _NET_WM_SYNC_REQUEST
WM_NORMAL_HINTS(WM_SIZE_HINTS):
        user specified location: 1012, 567
        user specified size: 186 by 220
        window gravity: NorthWest
```

**Comment by arielnmz** [comment]
So far I can use this rule to fix the popup menus:

    _NET_WM_WINDOW_TYPE:a *= '_KDE_NET_WM_WINDOW_TYPE_OVERRIDE'

But the call dialog still has that weird blur between the frame and the client content:

![image](https://user-images.githubusercontent.com/24735547/85957407-0c167580-b942-11ea-9ef3-6b946a1eab19.png)

**Comment by arielnmz** [comment]
No news on this?

**Comment by AdamantGarth** [comment]
I'm also having this issue. Thank you for a partial workaround though, @arielnmz.

Is there an issue comment or some documentation where it's explained why `picom` puts shadows on windows that other compositors do not? I never had an issue with shadows while using Mutter (and forks), Xfwm, Kwin or Compiz. If they're able to tell which windows shouldn't be decorated with shadows, `picom` should also be able to, right? No disrespect to developers, just genuinely curious.

**Reply by absolutelynothelix** [reply]
> Is there an issue comment or some documentation where it's explained why picom puts shadows on windows that other compositors do not? I never had an issue with shadows while using Mutter (and forks), Xfwm, Kwin or Compiz. If they're able to tell which windows shouldn't be decorated with shadows, picom should also be able to, right? No disrespect to developers, just genuinely curious.

i believe that picom aims to be configurable and not to make assumptions instead of an user.

**Reply by arielnmz** [reply]
The problem is that picom does not allow writing rules that match values within properties e.g. `_MOTIF_WM_HINTS[2]` to check for specific WM flags. But there is already a solution in the works for that: https://github.com/yshui/picom/issues/616

**Comment by OopsYao** [comment]
For me, adding `_GTK_FRAME_EXTENTS@:c` in the shadow and blur excluding rules fixes this, which comes from the sample config file.
https://github.com/yshui/picom/blob/0141bfb3a7424f365538c988dfb5b93d92a50d85/picom.sample.conf#L51
Picom version: vgit-98a5c.

**Comment by seqizz** [comment]
Sorry for necrobump, that extents thing doesn't work for the case sadly.

I realized I duplicated the same question here: https://github.com/yshui/picom/discussions/903 but so far no joy :)

**Comment by absolutelynothelix** [comment]
since this affects both telegram desktop windows and menus i'd do this:
```
shadow-exclude = [
  "class_g = 'TelegramDesktop'"
];

blur-background-exclude = [
  "class_g = 'TelegramDesktop'"
];
```

however, it's not a perfect solution.

**Reply by arielnmz** [reply]
There is a solution in the works to be able to write rules that match raw fields in properties. See: https://github.com/yshui/picom/issues/616 and https://github.com/tryone144/picom/commit/c349a92a5b2ff63d4dc31fbd0e9f3c2bced96068

Basically you would have to patch the code and rebuild picom and then you can use rules like:

    "_MOTIF_WM_HINTS[2]:32r = 0",

**Reply by absolutelynothelix** [reply]
ah, i saw your pull request on this. if you claim that it fixes the described issue and you've been using it for a long time without other issues then it'd be good to have it merged:
1) either update and clean up your pull request (if it needs to be);
2) or ask @tryone144 to make a pull request himself.

then we need to ask @yshui to review the pull request.

**Reply by arielnmz** [reply]
Yup, just asked @tryone144 to make a PR... one note though, I didn't claim it fixes this issue, only that the feature to access raw properties in the rules works fine. Even if that feature is added, the tricky part is coming up with the right rule for some windows. e.g. I haven't found a way of matching this transient window in a gtk3 client
![image](https://user-images.githubusercontent.com/24735547/231824086-f7bd762a-fe75-46be-9820-e5294c41257f.png)

**Reply by meliurwen** [reply]
@arielnmz for your GTK issue try these rules:

```txt
shadow-exclude = [
    ...
    "window_type = 'menu'",
    ...
];

fade-exclude = [
    ...
    "window_type = 'menu'"
    ...
]

blur-background-exclude = [
    ...
    "window_type = 'menu'",
    ...
];
```

They work for gtk4 apps menus, but I'm not sure for gtk3's since I've migrated quite some time ago and don't really remember; this is [my configuration](https://github.com/meliurwen/dotfiles/blob/master/picom/.config/picom/config) in case you find something useful.

Btw what is the name of that emoji application so I can try to reproduce it.

**Reply by absolutelynothelix** [reply]
@meliurwen, unfortunately, this disables effects for all menus. they play nice with some menus, that's why people want to disable effects only for menus belonging to a particular applications.

**Reply by arielnmz** [reply]
@meliurwen the application is smile https://flathub.org/apps/it.mijorus.smile and yes as @absolutelynothelix I think that rule disables it for all menus, but gtk2 menus work just fine, i wouldn't want to lose the pretty effect on them 😄
